File: 40-combination-sum-ii/combination-sum-ii.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        candidates.sort()
        cache = set()
        temp = []
        ans = []
        n = len(candidates)
        def solve(idx, total):
            if total ==target:
                cache.add(tuple(sorted(temp[:])))
                return
            
            for k in range(idx, n):
                if(k>idx  and candidates[k]==candidates[k-1]):
                    continue
                if total + candidates[k] > target:
                    break
                temp.append(candidates[k])
                solve(k+1, total + candidates[k])
                temp.pop()
        
        solve(0,0)
        unique_lists = [list(t) for t in cache]
        return unique_lists
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        

File: 40-combination-sum-ii/combination-sum-ii.java
class Solution {
    public void solve(int[] candidates, List<List<Integer>> ans,HashSet<List<Integer>> st, List<Integer> temp, int target, int idx){
        if(idx>=candidates.length){
            if(target==0 && !st.contains(temp)){
                ans.add(new ArrayList<>(temp));
                st.add(new ArrayList<>(temp));
            }
            return;
        }
        if(target>=candidates[idx]){
            temp.add(candidates[idx]);
            solve(candidates, ans, st, temp, target-candidates[idx], idx+1);
            temp.remove(temp.size()-1);
        }
        int next = idx+1;
        while(next<candidates.length && candidates[idx]==candidates[next]) next++;
        solve(candidates, ans, st, temp, target, next);
        return;
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> ans = new ArrayList<>();
        HashSet<List<Integer>> st = new HashSet<>();
        List<Integer> temp = new ArrayList<>();
        solve(candidates, ans, st, temp, target, 0);
        return ans;
    }
}